mlflow/haystack/autolog.py

```diff
@@ -0,0 +1,216 @@
+from opentelemetry.sdk.trace.export import SimpleSpanProcessor, SpanExporter
+from opentelemetry.trace import (
+    NoOpTracerProvider,
+    ProxyTracerProvider,
+    get_tracer_provider,
+    set_tracer_provider,
+)
+
+from mlflow.entities import LiveSpan, SpanType
+from mlflow.entities.span import create_mlflow_span
+from mlflow.tracing.constant import SpanAttributeKey, TokenUsageKey
+from mlflow.tracing.provider import _get_tracer
+from mlflow.tracing.trace_manager import InMemoryTraceManager
+from mlflow.tracing.utils import (
+    _bypass_attribute_guard,
+    generate_trace_id_v3,
+    get_mlflow_span_for_otel_span,
+)
+
+_logger = logging.getLogger(__name__)
+
+
+def setup_haystack_tracing():
+    from haystack import tracing as hs_tracing
+
+    hs_tracing.tracer.is_content_tracing_enabled = True
+
+    provider = get_tracer_provider()
+    hs_processor = HaystackSpanProcessor()
+    if isinstance(provider, (NoOpTracerProvider, ProxyTracerProvider)):
+        new_provider = SDKTracerProvider()
+        new_provider.add_span_processor(hs_processor)
+        set_tracer_provider(new_provider)
+    else:
+        if not any(
+            isinstance(p, HaystackSpanProcessor)
+            for p in provider._active_span_processor._span_processors
+        ):
+            provider.add_span_processor(hs_processor)
+
+    tracer = trace.get_tracer(__name__)
+    enable_tracing(OpenTelemetryTracer(tracer))
+
+
+def _infer_span_type_from_haystack(
+    comp_type: str | None,
+    comp_alias: str | None,
+    span: OTelReadableSpan,
+) -> SpanType:
+    s = (comp_type or comp_alias or span.name or "").lower()
+
+    if any(
+        k in s
+        for k in (
+            "llm",
+            "chat",
+            "generator",
+            "completion",
+            "textgen",
+            "chatgenerator",
+            "openai",
+            "anthropic",
+            "mistral",
+            "cohere",
+            "gemini",
+        )
+    ):
+        return SpanType.LLM
+
+    if "embedder" in s:
+        return SpanType.EMBEDDING
+
+    if "retriever" in s:
+        return SpanType.RETRIEVER
+
+    if "ranker" in s:
+        return SpanType.RERANKER
+
+    if "agent" in s:
+        return SpanType.AGENT
+
+    return SpanType.TOOL
+
+
+class HaystackSpanProcessor(SimpleSpanProcessor):
+    def __init__(self):
+        self.span_exporter = SpanExporter()
+        self._pipeline_io: dict[str, tuple[dict[str, Any], dict[str, Any]]] = {}
+
+    def on_start(self, span: OTelSpan, parent_context: Context | None = None):
+        tracer = _get_tracer(__name__)
+        tracer.span_processor.on_start(span, parent_context)
+
+        trace_id = generate_trace_id_v3(span)
+        mlflow_span = create_mlflow_span(span, trace_id)
+        InMemoryTraceManager.get_instance().register_span(mlflow_span)
+
+    def on_end(self, span: OTelReadableSpan) -> None:
+        mlflow_span = get_mlflow_span_for_otel_span(span)
```

Comment:
Thank you joel for the investigation. If the issue is caused because of the discrepancy between Otel span name and `mlflow_span._original_name`, is it possible to overwrite the otel span name before initializing mlflow span here? It would be clearer than overwriting the private field.